File: cycleInLL.cpp
// Time Complexity : O(n)
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No


// Your code here along with comments explaining your approach
// Floyd's algorithm
// 1. Initialize slow and fast pointers at start and move slow by one and fast by two
// 2. If slow and fast meet implies there is a cycle, else return nullptr
// 3. Start slow at head and fast at meeting point and advance both by one step, where they meet is the start of the cycle


/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode *detectCycle(ListNode *head) {
        if(!head)
            return 0;
        ListNode* slow=head;
        ListNode* fast=head;
        while(fast!=nullptr && fast->next!=nullptr){
            fast=fast->next->next;
            slow=slow->next;
            if(slow==fast)
                break;
        }
        if(!fast || !fast->next)
            return nullptr;
        slow=head;
        while(slow!=fast){
            fast=fast->next;
            slow=slow->next;
        }
        return slow;
    }
};

File: CycleInLL.py
# Time Complexity : O(n)
# Space Complexity : O(1)
# Did this code successfully run on Leetcode : Yes
# Any problem you faced while coding this : No


# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def detectCycle(self, head: Optional[ListNode]) -> Optional[ListNode]:
    # Let's take 2 pointers slow and fast slow moves by one and fast moves by two
        slow = head
        fast = head
        flag = False
    #There is a cycle if slow and fast meets at some point
        while fast is not None and fast.next is not None :
            slow = slow.next
            fast = fast.next.next
            if slow == fast:
                flag = True
                break
        if flag == False:
            return None
    #Move either of the fast or slow pointer to head and return the node where both points to same node
        fast = head
        while slow!=fast:
            fast = fast.next
            slow = slow.next
        return fast